### ISSUE TITLE
MDLX Parser: missing MDL format data

### ISSUE THREAD

**Issue by EzraZebra** [flowtsohg/mdx-m3-viewer#44]
There's some data which isn't represented in the MDL read/write.

For v800 models I'm sure you're aware of these:
- EventObject Global Sequence Id: this unfortunately causes Magos to be unable to open the model
```
	EventTrack [COUNT excluding globseqid] {
		GlobalSeqId [ID],
		[TRACKS...],
	}
```
- Model Animationfile: in practice this is useless, but might as well put it in there
```
	AnimationFile "[PATH]",
```

**Comment by EzraZebra** [comment]
- Add to model.ts `loadModelBlock`:
```
      } else if (token === 'AnimationFile') {
        this.animationFile = stream.readSafe();
      }
```
- Add to model.ts `saveModelBlock`:
```
    stream.writeStringAttrib('AnimationFile', this.animationFile);
```

**Comment by flowtsohg** [comment]
Do you have models with the relevant v800 properties that are readable by the game?

As to Reforged, as far as I am aware it doesn't have MDL.
While it would be nice to have the data stay there if you are messing with the formats, I only want things that the game can actually load.
Unless those names and values are actually supported by the game?

**Comment by EzraZebra** [comment]
I don't have Reforged so unfortunately I can't test it.
AFAIK Retera's Model Studio uses native dlls to convert models, so I assumed these were native MDL format tokens. But I guess those custom then? I'll contact him about it.

**Comment by flowtsohg** [comment]
The v800 things look correct (as in they are missing in my code), it's just preferable to actually have data to test it with. If you don't have models it's fine, I can probably find one in my local mess of wc3 resources that has these properties.

As for native dlls, I am only aware of Storm.dll being used for MPQ, although I think that isn't being used (or is even available?) either with newer versions.

**Comment by Retera** [comment]
Retera's Model Studio is using fan-made java code to convert models. That java code converts to an invented MDL format that is not the real format. For example, there are not 2 color chunks. This is because I didn't fix my code yet and use Blizzard Reforged MDL format.

See this:
https://www.hiveworkshop.com/threads/retera-model-studio-reforged-hack.320070/page-2#post-3387392

For example, there are not two color chunks. I did this in a hurry during beta because I had no idea how to parse the binary data. These chunks are actually (quoted from the above):
```
    static LifeSpan %.3f,                                     // or float track
    static EmissionRate %.3f,                                 // or float track
    static Speed %.3f,                                        // or float track
    static Color { %.3f, %.3f, %.3f },                        // or float3 track
    static Alpha %.3f,                                        // or float track
    ReplaceableId %d,                                         // if not 0
```

Likewise, "FlagString" is misnamed, and is actually "AnimVisibilityGuide"

**Comment by EzraZebra** [comment]
I see, thanks for the info! Pretty naive of me to assume it was all legit blizzard format :P

As for the AnimationFile and eventobject GlobalSeqId, just tested it and appears to work.
[herotinker2.zip](https://github.com/flowtsohg/mdx-m3-viewer/files/4567983/herotinker2.zip)

**Comment by flowtsohg** [comment]
Hopefully fixed the v800 stuff, thanks for the test model.

That dump does look legit, but I don't have Reforged either to test any of it.

**Comment by EzraZebra** [comment]
@Retera 
[murgulreaver.mdla28712.txt](https://github.com/flowtsohg/mdx-m3-viewer/files/4593588/murgulreaver.mdla28712.txt)

So apparently there's a MDL file in the Reforged CASC at:
 "1.32.x_hd.w3mod\units\creeps\murgulreaver\murgulreaver.mdla28712"

eejin sent it to me, and these are the same as in that MDL dump:
Tangents
Skinweights
LOD/LOD Name
BindPose (although the spacing is aligned, but I doubt that impacts parsing)

I think we can be reasonably confident that the rest of the MDL dump is accurate as well.

**Comment by flowtsohg** [comment]
I have 4 problems with it:
1) The way triangles are formatted is different.
2) Popcorn emitters have no static alpha in MDX, but have in MDL?
3) FaceFX has no options string, but I suppose it could be set to some default value.
4) I can't test any of these things.

**Comment by flowtsohg** [comment]
The static alpha is just read in my code as a part of the RGBA static color...DUH 🤦

**Comment by EzraZebra** [comment]
> The way triangles are formatted is different (does it now support both? is it now version dependent? is there any reason to needlessly add tokens between the triangles--oh wait this is Blizzard)

Huh, didnt notice that, only looked at the v900 stuff. Will have a look later. I wonder if that will work in classic, probably not.

> FaceFX has no options string, but I suppose it could be set to some default value.
I can't test any of these things.

I guess you're talking about the dump here? Didnt see those in the attached MDL. But afaik facefx only have a name and a target/path? Both are represented in the dump.
Unfortunately I can't test it either for now. I may ask in the hive discord if someone else can test it.

And yeah the static alpha is from the rgba, didnt get that immediately either xD

**Comment by flowtsohg** [comment]
Retera confirmed that the different triangle formatting is actually not related to Reforged, MDL just allows you to split the indices into vectors.
I don't know if it has to be the size of the primitive, like 3 for triangles, or if it's completely arbitrary.
I updated the reading code to support this, which I'll commit at some point, but I didn't change the writing code in any way, since I see no purpose in adding more useless characters to the output.

The FaceFX part is unclear to me. The dump shows a string name before the block, but I highly doubt that is the options string.
No idea why there's NumFaceFX, perhaps it's related to the fact there are `.facefx` and `.facefx_ingame` files, but then why aren't both (optional) paths in one block.

**Comment by EzraZebra** [comment]
Interesting about the triangles. Could be used to improve readability for MDL files (for my converter tool).

About the facefx, the options string is the Target part, right? I've only seen `"Node"` as its value, so couldn't it be that string before the block? Since facefx have two strings and the MDL dump also has two strings, it seems reasonable to me.

**Comment by flowtsohg** [comment]
I forgot it's the shorter string, confused it with popcorn's visibility guide thing.
Perhaps the idea was to have multiple face effects with different targets, although I have no idea what the target actually means.
Technically also MDX could have this by simply having multiple FAFX chunks, and I wouldn't even notice because my parser wouldn't care and just end up with the last one, need to look into this.

Either way, these new MDL tokens and values are legit.

**Comment by EzraZebra** [comment]
In my parser I've put the facefx in a vector<array<string, 2>>. It will probably pretty much always be a vector of size 1, but just in case.  
Now that I think about it, maybe it's for models like mortar team and two headed ogre, which actually have two faces?

**Comment by flowtsohg** [comment]
That's a pretty funny idea, however those models and some others I've looked at have just one.
Maybe it's something that we are missing, or something that isn't used yet, or blizz not knowing how to write formats, for now who knows 🤷 

I have it all implemented already, so I'll close this issue once I commit.

**Comment by EzraZebra** [comment]
I see, I guess that's not it then :thinking: 

BTW, I just solved a major in my parser, turns out it was caused by not reading the number of geoset face into the faceGroups variable. This causes a gamecrashing error when converting from MDL to MDX.
Should I make a new issue or is it solved already? (since you added support for groups of three faces in the mdl reader)

**Comment by flowtsohg** [comment]
Ah yeah that's missing indeed.

**Comment by EzraZebra** [comment]
Popcorn emitter is missing MDL read/write for flags (SortPrimsFarZ `0x10000`, Unshaded `0x8000`, Unfogged `0x40000`).

**Comment by flowtsohg** [comment]
In general I don't trust many of these flags, including the object type ones, and including older objects like the different emitters, but I don't care enough right now to start checking tons of models.

**Comment by flowtsohg** [comment]
It looks to me like Layer's EmissiveGain should actually be named Emissive.
Probably same for the animation(?).
Can this be confirmed? @Retera

**Comment by flowtsohg** [comment]
Confirmed to be fake tokens.

**Comment by flowtsohg** [comment]
N face effects are indeed supported.
For MDX, they all come in the one FAFX chunk one after the other like sequences and textures, no explicit count, rather calculated by dividing the chunk size with the size of an effect, i.e. 340 bytes.
For MDL, it's not tested yet, but I assume every FAFX block is an object, loaded normally like geosets and nodes.

**Comment by flowtsohg** [comment]
Implemented N face effects in d1d182ff29be2cf349b3e3b5576aab4b2ab96056.
